API_MANIFESTO.md

```diff
@@ -0,0 +1,91 @@
+Dependency with implementations will be **logical**. This will keep the modularity needed for the different use cases.
+
+Examples:
+ * Cache use case
+ 
+ * NoSQL data store use case
+ 
+ * Tasks, concurrency, counters uses cases
+
+The dependencies to thirty-party libraries must be avoided. This module should contain dependencies to JDK classes, 
+not more.
+
+# User First
+
+* Before making an API public, consider to test the usability with someone not familiar with the feature
+
+* Convention over Configuration: Make sure a default behaviour exists. Don't make me think.
+
+* Avoid exposing internals as much as possible. Even if your feature lays on top of a cache today, internals can change.
+
+* Each public API/Feature must have a simple tutorial showcasing the example. Without a demo, the API does not exist
+
+* If you are unsure about an API, mark it as ```@Experimental```. 
+
+* Don't let ```@Experimental``` tag last forever. 
+
+* Ask for user's feedback actively (zulip, user forums, blog post, twitter, github ...)
+
+* Be clear about whats PUBLIC API and what internals/implementations are. 
+
+
+# Client/Server First
+
+* New interfaces should be the same for embedded and client/server
+
+* Always start defining your API considering users in client/server mode. Embedded implementation will follow.
+
+* If a particular need is demanded for embedded, before doing something specific
+   - Discuss with the team the need. Is it really mandatory?
+   - Is this a general purpose or a really advanced need? 
+   - Will the functionality be widely used? Why is this embedded only then?
+   
+Embedded only API should be an exception at this point and they might not live in the **infinispan-api** module.
+
+
+# When do we create a new module/dependency?
+
+This has to be discussed case by case with the team. Several factors need to be considered.
+
+* Can the implementation of a new API live in an already existing module?
+
+* Would be better for the user if we create a new module, but we add the needed dependencies for the user?
+
+
+# Creating API
+
+* When you create a new API, you should consider the use case it's about to solve. 
+
+* Avoid Swiss knifes on interfaces. Keep it focuses and small enough.
+
+* Remember public APIs are like diamonds, they last forever. Rarely we will be allowed to just remove an API without 
+breaking compatibility. 
+
+
+# Evolving API
+
+* Adding a new method to an existing interface must be done with care. Consider the need of the method versus creating a
+new interface.
+
+* If there is no other choice that adding a new method, provide a **default implementation**
+
+* Methods tend to need parameter overloading overtime. Use a **Parameters** object to assure evolution over time instead 
+of a list of parameters
+
+
+# Deprecating API
+
+* Deprecating a public method is sad, but it happens. Provide the proper documentation and the reasons for the deprecation.
+Add the alternative code explaining the changes. 
```

Comment:
I think here we should say what *proper* documentation is, i.e. updating the upgrading doc, Javadocs, User guide etc.

@oraNod any thoughts on this Manifesto in general and the docs impact?

Reply:
@ryanemerson I think the idea of the Manifesto is a good one. It's useful to have a set of principles like this to guide the work. As for docs impact, I'd like to expand a bit under the "User First" section because the focus of documentation is the user.

Agreed that we should define/expand "proper documentation" to be more specific.

Reply:
I'm adding a section for "Documentation" in the doc. More than welcome for whatever you want to add there @oraNod ! Or any other place in this file